colbert/index_updater.py

```diff
@@ -0,0 +1,178 @@
+from colbert.indexing.collection_encoder import CollectionEncoder
+
+from colbert.utils.utils import lengths2offsets, print_message, dotdict, flatten
+from colbert.indexing.codecs.residual import ResidualCodec
+from colbert.indexing.utils import optimize_ivf
+from colbert.search.strided_tensor import StridedTensor
+from colbert.modeling.checkpoint import Checkpoint
+from colbert.utils.utils import print_message, batch
+from colbert.data import Collection
+from colbert.indexing.codecs.residual_embeddings import ResidualEmbeddings
+from colbert.indexing.utils import optimize_ivf
+
+class IndexUpdater:
+    
+    def __init__(self, config, searcher, checkpoint=None):
+        self.config = config
+        self.searcher = searcher
+        if checkpoint:
+            self.checkpoint = Checkpoint(checkpoint, config)
+            self.encoder = CollectionEncoder(config, self.checkpoint)
+        self.index_path = searcher.index
+        self._load_disk_ivf()
+        
+        
+    def remove(self, pids, persist_to_disk=False):
+        for pid in pids:
+            print(f"removing {pid}...")
+            self._remove_pid_from_ivf(pid, persist_to_disk)
+            if persist_to_disk:
+                self._load_metadata()
+                self._remove_passage_from_disk(pid)
+    
+    def add(self, pids, persist_to_disk=False):
+        if not self.checkpoint:
+            raise ValueError("No checkpoint was provided at IndexUpdater initialization.")
+        # TODO: implement the add-passage functionalities
+        
+    def _load_disk_ivf(self):
+        print_message(f"#> Loading IVF...")
+
+        if os.path.exists(os.path.join(self.index_path, "ivf.pid.pt")):
+            ivf, ivf_lengths = torch.load(os.path.join(self.index_path, "ivf.pid.pt"), map_location='cpu')
+        else:
+            assert os.path.exists(os.path.join(self.index_path, "ivf.pt"))
+            ivf, ivf_lengths = torch.load(os.path.join(self.index_path, "ivf.pt"), map_location='cpu')
+            ivf, ivf_lengths = optimize_ivf(ivf, ivf_lengths, self.index_path)
+
+        self.curr_ivf = ivf
+        self.curr_ivf_lengths = ivf_lengths
+        
+    def _remove_pid_from_ivf(self, pid, persist_to_disk):
+        new_ivf = []
+        new_ivf_lengths = []
+        runner = 0
+        for l in self.curr_ivf_lengths.tolist():
+            num_removed = 0
+            for i in range(runner, runner + l):
+                if self.curr_ivf[i] != pid:
+                    new_ivf.append(self.curr_ivf[i])
+                else:
+                    num_removed += 1
+            runner += l
+            new_ivf_lengths.append(l - num_removed)
+            
+        assert runner == len(self.curr_ivf.tolist())
+        assert sum(new_ivf_lengths) == len(new_ivf)
+        
+        new_ivf = torch.tensor(new_ivf)
+        new_ivf_lengths = torch.tensor(new_ivf_lengths)
+        
+        new_ivf_tensor = StridedTensor(new_ivf, new_ivf_lengths, use_gpu=False)
```

Comment:
We're currently constructing a new `StridedTensor` every time because I'm not sure whether there's a good way of updating the existing `StridedTensor` which computes some metadata in its constructor - @okhat does this make sense to you or is there a better option?

Reply:
It's probably okay (for now at least), as long as we allow adding/removing a batch all at once.

Reply:
Yes, just made changes to ensure it's one reconstruction per batch!